teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,997 @@
+    a .Internally, 
+      i. `$(step.results.resultName.path) = /tekton/steps/<step-name>/results/resultName`
+`Task` authors should be able to:
+1. fetch results from underlying `Steps` using  `$(step.stepName.results.resultName)`.
+2. For backwards compatibility, they can write inlined `Steps` with results written to `$(results.resultName.path)`. These results will automatically be surfaced to the `Task` by the entrypointer.
+3. However, if they write inlined `Steps` where `Results` are written to `$(step.results.resultName.path)` they need to fetch the `Step` `Results` using `$(step.stepName.results.resultName)`.
+
+<table>
+<tr>
+<td>
+
+```yaml
+# This is how a step author would 
+# declare the results that it needs.
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  results:
+    - name: digest
+      description: “One result”
+  image: alpine
+  script: |
+    echo DIGEST >> $(step.results.digest.path)
+```
+</td>
+<td rowspan=2>
+
+```yaml
+# This is how a task author would 
+# provide the results that the step needs.
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  results:
+    - name: digest1
+      value: $(steps.step1.results.digest)
+    - name: digest2
+      value: $(steps.step2.results.digest)
+    - name: normalResult
+      description: “Normal result.”#Automatically sent 
+# to the TaskRun CRD. No need to fetch it via 
+#values since it was written to the old path.
+  steps:
+    - name: step1
+      stepRef: 
+        name: myStep
+    - name: step2
+      stepRef:
+        name: myOtherStep
+    - name: step3
+      image: foo
+      script: |
+        echo RESULT >> $(results.normalResult.path)
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myOtherStep
+spec:
+  results:
+    - name: digest
+      description: “Another result”
+  image: alpine
+  script: |
+    echo OTHER-DIGEST >> $(step.results.digest.path)
+```
+</td>
+</tr>
+</table>
+
+Currently, `Steps` cannot access `Results` from previous `Steps` via variable interpolation. `Steps` are currently inlined only so `Task` authors can write logic to read the `Result` files written by previous `Steps`. 
+
+However, since the `StepAction` are defined in isolation and are referenced in `Tasks`, it is challenging for the `Task` authors to pass `Results` from one `Step` to another without Tekton providing some API. Therefore, we should enable referencing of `Results` produced by `Steps` via `Parameters` of the subsequent `Steps`. 
+
+`Task` authors should be able to extract `Results` from previous `Steps` (i.e. written to  `$(step.results.resultName.path)`) using `$(steps.stepName.results.resultName)`. `Results` written to `$(results.resultName.path)` are not referenceable.
```

Comment:
I am not sure I follow this part. Do we want to enable referencing results from one step to the other ? Only for `StepAction` or any steps ? Any reason to do this instead of passing information through files (aka give the `path` of the Results to a `StepAction` param for it to know what to read) ?

Reply:
I think any Steps but here, I'm proposing that the reference to that result should be done via `step.results.resultName.path` instead of `result.resultName.path`. During referencing in a later step via `$(steps.StepName.results.resultName)`, the entrypointer does not know which step to reference it from. The reason for this step referencing is because multiple step actions in a Task could produce a result name same thing (e.g. `digest`, which is a common name) but actually need to be different. The Task author putting together these steps should be able to capture their results without the risk of them overwriting each other.

Reply:
> Do we want to enable referencing results from one step to the other ?

The proposal is to enable referencing results produced by any previous step.

> Only for StepAction or any steps ? 

only for StepAction. `StepAction` is mostly envisioned as a replacement of a `task` with a single step.

>  Any reason to do this instead of passing information through files (aka give the path of the Results to a StepAction param for it to know what to read) ?

It's the same concept, some content is written to a file in a pre-defined location in `/tekton/steps/<step-name>` v/s user specifies their own choice of a path. The users still have a choice to not rely on `results` but create a file as a means of information exchange.

Reply:
> > Do we want to enable referencing results from one step to the other ?
> 
> The proposal is to enable referencing results produced by any previous step.
> 
Yes, here I intended between steps. That's because when users use `StepActions`, they are using what someone else has already written. When inlining steps, it's a very different behaviour because the same author is writing their custom  script to handle results created by previous steps. 

> > Only for StepAction or any steps ?
> 
> only for StepAction. `StepAction` is mostly envisioned as a replacement of a `task` with a single step.
> 

I don't think we can distinguish between Step and StepAction once the Pod is created because the StepAction is basically converted to an inlined step after it is fetched from the cluster/remote resource. Therefore, this is meant for any Steps.

> > Any reason to do this instead of passing information through files (aka give the path of the Results to a StepAction param for it to know what to read) ?
> 
> It's the same concept, some content is written to a file in a pre-defined location in `/tekton/steps/<step-name>` v/s user specifies their own choice of a path. The users still have a choice to not rely on `results` but create a file as a means of information exchange.

Agreed. The StepAction could totally accept a path to a result file as a param that it could read. However, I think we have the full ability to do this automatically via the entryPointer so it might be best to just provide this as an API sugar so that users don't have to worry about unmarshaling the results. Normally this wouldn't even be required when just inlining the steps since the users have the ability to work around it. Since the StepActions are written in isolation, this is more needed now so that Task authors putting multiple step actions together can rely on passing results to following steps.I think, adding this capability will increase the chances for adoption since it will reduce UX barrier.

Reply:
> Agreed. The StepAction could totally accept a path to a result file as a param that it could read. However, I think we have the full ability to do this automatically via the entryPointer so it might be best to just provide this as an API sugar so that users don't have to worry about unmarshaling the results. Normally this wouldn't even be required when just inlining the steps since the users have the ability to work around it. Since the StepActions are written in isolation, this is more needed now so that Task authors putting multiple step actions together can rely on passing results to following steps.I think, adding this capability will increase the chances for adoption since it will reduce UX barrier.

I am not sure, I follow here. The proposal is to be able to use a `result` as "content" for a parameter value right ? It's not that different from having a parameter that says "read the data from that file", but it is suger 🙃. The "funny" use case would be, some refers to a `results` — so getting the content of it — to write a file with it to pass to the underlying process. That use case would have been better supported with just giving the path (which is still possible so 🧑‍🔧 ).

IMO, this is a "nice to have" feature but is not necessarily a requirement. But 👍🏼 anyway 😝 .

Reply:
Agreed that this is not a requirement but a nice to have. I think when I initially put together the tasks-in-tasks doc, some readers even explicitly asked if this type of feature would be available.